File: apps/mobile/lib/models/store.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// Dart mirror of `packages/contracts/src/store.ts`.
///
/// Kept in sync by hand — Dart cannot import the TypeScript contract. Any
/// change there must be reflected here. Only the fields the dealer app reads
/// are modelled; the rest stay untouched on the document.
enum StoreStatus {
  pending,
  approved,
  rejected,
  suspended;

  static StoreStatus parse(String? raw) => switch (raw) {
        'approved' => StoreStatus.approved,
        'rejected' => StoreStatus.rejected,
        'suspended' => StoreStatus.suspended,
        _ => StoreStatus.pending,
      };

  bool get canPublish => this == StoreStatus.approved;
}

/// Days a lapsed plan keeps its listings live. Mirrors
/// SUBSCRIPTION_GRACE_DAYS in packages/contracts/src/constants.ts.
const int kSubscriptionGraceDays = 7;

class Subscription {
  const Subscription({
    required this.plan,
    required this.status,
    this.expiresAt,
    this.graceEndsAt,
  });

  final String plan; // free | monthly | yearly
  /// As last written by the backend sweep. Prefer [statusNow], which is exact.
  final String status; // none | active | grace | expired
  final DateTime? expiresAt;
  final DateTime? graceEndsAt;

  /// The status right now, derived from the dates rather than read.
  ///
  /// `status` is maintained by an hourly Cloud Function, so between a plan
  /// lapsing and the next sweep the stored value still says `active`. Deriving
  /// it here keeps the dealer's screen honest in that window — otherwise the
  /// app cheerfully reports an active plan for up to an hour after it ended,
  /// and then appears to change its mind on its own.
  ///
  /// Falls back to the stored value when the dates are missing, which is the
  /// case for stores created before those fields existed.
  String statusNow([DateTime? now]) {
    if (plan == 'free') return 'none';
    final expiry = expiresAt;
    if (expiry == null) return status;

    final at = now ?? DateTime.now();
    if (at.isBefore(expiry)) return 'active';

    final graceEnd =
        graceEndsAt ?? expiry.add(const Duration(days: kSubscriptionGraceDays));
    return at.isBefore(graceEnd) ? 'grace' : 'expired';
  }

  bool isPaid([DateTime? now]) {
    final s = statusNow(now);
    return s == 'active' || s == 'grace';
  }

  /// Lapsed but still honoured — listings stay live through the grace window.
  bool inGrace([DateTime? now]) => statusNow(now) == 'grace';

  /// Lapsed and no longer honoured. The backend has dropped this store back to
  /// the free allowance, so active listings above 10 have been unpublished.
  bool hasExpired([DateTime? now]) => statusNow(now) == 'expired';

  /// Whole days until [graceEndsAt], for the countdown a dealer in grace needs.
  int? graceDaysLeft([DateTime? now]) {
    if (statusNow(now) != 'grace') return null;
    final end = graceEndsAt ??
        expiresAt?.add(const Duration(days: kSubscriptionGraceDays));
    if (end == null) return null;
    final remaining = end.difference(now ?? DateTime.now());
    if (remaining.isNegative) return 0;
    // Ceiling, so a plan with two hours left reads as "1 day" rather than "0",
    // and exactly seven days reads as 7 rather than 8. `inHours ~/ 24 + 1`
    // looks like the same thing and is wrong on every exact boundary.
    return (remaining.inMilliseconds / Duration.millisecondsPerDay).ceil();
  }

  factory Subscription.fromMap(Map<String, dynamic>? m) => Subscription(
        plan: (m?['plan'] as String?) ?? 'free',
        status: (m?['status'] as String?) ?? 'none',
        expiresAt: (m?['expiresAt'] as Timestamp?)?.toDate(),
        graceEndsAt: (m?['graceEndsAt'] as Timestamp?)?.toDate(),
      );
}

class Store {
  const Store({
    required this.storeId,
    required this.businessName,
    required this.ownerName,
    required this.phone,
    required this.whatsapp,
    required this.cacNumber,
    required this.address,
    required this.state,
    required this.city,
    required this.description,
    required this.slug,
    required this.status,
    required this.visible,
    required this.activeListingCount,
    required this.subscription,
    this.email = '',
    this.landmark = '',
    this.automotiveCategory = '',
    this.rejectionReason,
  });

  final String storeId;
  final String businessName;
  final String ownerName;
  final String phone;
  final String whatsapp;
  final String cacNumber;
  final String address;
  final String state;
  final String city;
  final String description;

  /// From the client-approved registration design rather than the SOW §2 field
  /// list. Default to '' on read: stores registered before these existed have
  /// no such key, and a missing field must not break the dealer app.
  final String email;
  final String landmark;
  final String automotiveCategory;

  // Backend-controlled — read-only here by design (ADR-001 #4).
  final String slug;
  final StoreStatus status;
  final bool visible;
  final int activeListingCount;
  final Subscription subscription;
  final String? rejectionReason;

  /// Mirrors activeLimitFor() in functions/src/publishListing.ts.
  /// Display only — the authoritative check happens in the transaction.
  int get activeListingLimit => subscription.isPaid() ? 200 : 10;

  bool get atListingLimit => activeListingCount >= activeListingLimit;

  factory Store.fromDoc(DocumentSnapshot<Map<String, dynamic>> doc) {
    final d = doc.data() ?? <String, dynamic>{};
    return Store(
      storeId: doc.id,
      businessName: (d['businessName'] as String?) ?? '',
      ownerName: (d['ownerName'] as String?) ?? '',
      phone: (d['phone'] as String?) ?? '',
      whatsapp: (d['whatsapp'] as String?) ?? '',
      cacNumber: (d['cacNumber'] as String?) ?? '',
      address: (d['address'] as String?) ?? '',
      state: (d['state'] as String?) ?? '',
      city: (d['city'] as String?) ?? '',
      description: (d['description'] as String?) ?? '',
      email: (d['email'] as String?) ?? '',
      landmark: (d['landmark'] as String?) ?? '',
      automotiveCategory: (d['automotiveCategory'] as String?) ?? '',
      slug: (d['slug'] as String?) ?? '',
      status: StoreStatus.parse(d['status'] as String?),
      visible: (d['visible'] as bool?) ?? false,
      activeListingCount: (d['activeListingCount'] as num?)?.toInt() ?? 0,
      subscription: Subscription.fromMap(d['subscription'] as Map<String, dynamic>?),
      rejectionReason: d['rejectionReason'] as String?,
    );
  }
}

File: packages/contracts/src/store.ts
import type { Timestamp } from './common';

/**
 * A dealer business. Document ID is always the dealer's Firebase Auth uid,
 * so ownership checks in security rules reduce to `uid == storeId`.
 *
 * Collection: `stores/{storeId}`
 */

/** SOW §3: registration lifecycle, driven by the admin portal. */
export type StoreStatus = 'pending' | 'approved' | 'rejected' | 'suspended';

export type SubscriptionPlan = 'free' | 'monthly' | 'yearly';

/**
 * `grace` = paid period lapsed but listings are still live (see
 * SUBSCRIPTION_GRACE_DAYS). `expired` = lapsed and listings auto-unpublished
 * back down to the free limit.
 */
export type SubscriptionStatus = 'none' | 'active' | 'grace' | 'expired';

export interface Subscription {
  plan: SubscriptionPlan;
  status: SubscriptionStatus;
  startedAt: Timestamp | null;
  expiresAt: Timestamp | null;
  graceEndsAt: Timestamp | null;
  /** Paystack reference of the transaction that activated this period. */
  lastPaymentReference: string | null;
}

/** Fields the dealer submits at registration and may edit thereafter. SOW §2. */
export interface StoreProfileInput {
  businessName: string;
  ownerName: string;
  /** E.164, e.g. +2348031234567. Sourced from the authenticated phone number. */
  phone: string;
  whatsapp: string;
  /** Corporate Affairs Commission number. Verified manually by admin (SOW §3). */
  cacNumber: string;
  address: string;
  state: string;
  city: string;
  description: string;

  /**
   * The three fields below come from the client-approved registration design
   * rather than the SOW §2 field list. They are dealer-owned contact and
   * classification data with no security consequence, so they live alongside
   * the rest of the profile rather than behind a callable.
   *
   * Optional on read: stores registered before this was added will not have
   * them, and a missing field must not break the dealer app.
   */

  /** Business email. Dealers still authenticate by phone — this is contact only. */
  email?: string;
  /** Nearest landmark, e.g. "Opposite Ladipo Main Gate". Aids buyers finding a stall. */
  landmark?: string;
  /** Which automotive vertical the shop trades in. See AUTOMOTIVE_CATEGORIES. */
  automotiveCategory?: string;
}

export interface Store extends StoreProfileInput {
  storeId: string;

  // --- Backend-controlled below this line ---------------------------------
  // Dealers have read-only access. Enforced by security rules; see security.ts.

  /** Public URL segment: naijapartshub.ng/store/{slug}. SOW §6. */
  slug: string;
  status: StoreStatus;
  /** Set by admin when status is 'rejected'. Surfaced to the dealer. */
  rejectionReason: string | null;
  /** SOW §3: "Control whether a store is publicly visible". */
  visible: boolean;
  /** Maintained transactionally by publishListing. Never trust a client value. */
  activeListingCount: number;
  subscription: Subscription;

  termsAcceptedAt: Timestamp | null;
  createdAt: Timestamp;
  updatedAt: Timestamp;
  approvedAt: Timestamp | null;
  /** Admin uid that last approved or rejected this store. */
  reviewedBy: string | null;
}

/**
 * Slug reservation. Firestore has no unique constraint, so uniqueness is
 * enforced by writing this document inside the same transaction that assigns
 * the slug: the create fails if the ID already exists.
 *
 * Collection: `storeSlugs/{slug}`
 */
export interface StoreSlugReservation {
  storeId: string;
  createdAt: Timestamp;
}
